### PR TITLE
Add smmla/ummla support in quantized Conv2d

### PR DESCRIPTION
**Pull request by giuseros**
### High level description of the submission
This introduces support for `smmla`/`ummla` instructions in TVM:
- Added `is_mmla_available` function in `arm_utils.py`
- Added the tiling node + tensorization schedule in `conv2d_gemm.py`
- Added the intrinsic support in `tensor_intrin.py`
- Added the test-case in `test_topi_conv2d_int8.py`

### RFC
This PR is based on the following RFC: https://discuss.tvm.apache.org/t/rfc-improve-quantized-convolution-through-mmla-instructions/8336

Change-Id: Iff48c77f16fe1e64ecb733da965a879651ce635f

### CODE REVIEW

**Comment by giuseros** [comment]
cc: @anijain2305 , @FrozenGene , @mbaret , @u99127

**Review by cbalint13 — APPROVED** [review]
Nice addition @giuseros !
LGTM

File: python/tvm/topi/arm_cpu/arm_utils.py [hunk 1]
```diff
@@ -49,6 +49,15 @@ def is_dotprod_available():
     return arch_version >= 8.4 or ((arch_version in (8.2, 8.3)) and "+dotprod" in target.mattr)
 
 
+def is_mmla_available():
+    """ Checks whether the hardware has support for fast Int8 arithmetic operations. """
```
Comment (mbaret) [inline python/tvm/topi/arm_cpu/arm_utils.py:53]: I think this should be explicit

File: python/tvm/topi/arm_cpu/arm_utils.py [hunk 2]
```diff
@@ -77,7 +86,10 @@ def get_tiling_B_interleaved_t(interleave_A):
     tile_rows_B: the output tile rows of B'
     tile_cols_B: the output tile columns of B'
     """
-    if is_dotprod_available():
+    if is_mmla_available():
+        tile_rows_B = 12
+        tile_cols_B = 8
```
Comment (mbaret) [inline python/tvm/topi/arm_cpu/arm_utils.py:91]: Explain the numbers like below.

File: python/tvm/topi/arm_cpu/conv2d_gemm.py [hunk 3]
```diff
@@ -134,7 +135,10 @@ def compute_conv2d_gemm_without_weight_transform(
     #   - Conv2DGemmWeightTransformRel in src/relay/op/nn/convolution.h
     # In order to have more information
     #
-    if is_dotprod_available() and interleave_A:
+    if is_mmla_available():
+        tile_rows_A = 8
+        tile_cols_A = 8
```
Comment (mbaret) [inline python/tvm/topi/arm_cpu/conv2d_gemm.py:140]: Explain

File: python/tvm/topi/arm_cpu/conv2d_gemm.py [hunk 4]
```diff
@@ -271,7 +322,17 @@ def schedule_conv2d_gemm_interleaved(cfg, s, out, final_out):
 
     k = C_interleaved.op.reduce_axis[0]
     _, M, N = C.shape
-    if is_dotprod_available():
+    if is_mmla_available():
+        gemm_acc = gemm_acc_2x2_int8_int8_int32(in_type)
+        xi_inner, yi_inner = C_interleaved.op.axis[5:7]
```
Comment (mbaret) [inline python/tvm/topi/arm_cpu/conv2d_gemm.py:327]: [-2:]?

File: python/tvm/topi/arm_cpu/tensor_intrin.py [hunk 5]
```diff
@@ -879,6 +879,105 @@ def _instr(index):
     )
 
 
+def gemm_acc_2x2_int8_int8_int32(dtype):
+    """
+    Int8 2x2 matrix multiplication using smmla/ummla instructions
+    This function takes two arrays of int8 datatype -- A[2][8] and
+    B[2][8] and produces a 2x2 matrix which is equal to A*B
```
Comment (mbaret) [inline python/tvm/topi/arm_cpu/tensor_intrin.py:886]: A*(B transposed)

File: python/tvm/topi/arm_cpu/tensor_intrin.py [hunk 6]
```diff
@@ -879,6 +879,105 @@ def _instr(index):
...
+            for (int i = 0; i < 2; i++){
+                for (int j = 0; i < 2; i++){
+                    for (int k = 0; k < 8; k++){
+                        C[i][j] += A[i][k] * B[j][k]
+                    }
+            }
+        }
+
+    Notes:
+        * The rows of matrix B are transposed
```
Comment (mbaret) [inline python/tvm/topi/arm_cpu/tensor_intrin.py:901]: Matrix B isn't transposed, at least not the B defined here.

File: python/tvm/topi/arm_cpu/tensor_intrin.py [hunk 7]
```diff
@@ -879,6 +879,105 @@ def _instr(index):
...
+        A.shape, dtype, name="aa_buffer", offset_factor=1, strides=[te.var("sa"), 1]
+    )
+    bb_buffer = tvm.tir.decl_buffer(
+        B.shape, dtype, name="bb_buffer", offset_factor=1, strides=[te.var("sb"), 1]
+    )
+    cc_buffer = tvm.tir.decl_buffer(
+        C.shape, dtype="int32", name="cc_buffer", offset_factor=1, strides=[te.var("sc"), 1]
+    )
+
+    llvm_intrin = "llvm.aarch64.neon.smmla" if dtype == "int8" else "llvm.aarch64.neon.ummla"
```
Comment (mbaret) [inline python/tvm/topi/arm_cpu/tensor_intrin.py:934]: Just check that say a float32 can't get through to here.

**Review by mbaret — CHANGES_REQUESTED** [review]

File: python/tvm/topi/arm_cpu/tensor_intrin.py (hunk 6 above)
Comment (giuseros) [reply python/tvm/topi/arm_cpu/tensor_intrin.py]: I agree that the comment is misleading. I removed it, and back-propagated this to the other intrinsics

File: python/tvm/topi/arm_cpu/tensor_intrin.py (hunk 7 above)
Comment (giuseros) [reply python/tvm/topi/arm_cpu/tensor_intrin.py:933]: I am checking now at tensorization time, and asserting when I instantiate the intrinsic

**Comment by giuseros** [comment]
Thanks for the review. The only caveat here is that we would need support for LLVM-11 in the CI to run the integration test (llvm-8 does not have i8mm support). I am commenting out the test for now, and I will enable it back once the LLVM in the CI will be updated

**Review by mbaret — APPROVED** [review]
LGTM, nice job again Giuseppe.

**Comment by giuseros** [comment]
Hi @anijain2305 , @FrozenGene , 

Any chance you could review this? 

Thanks in advance!

**Review by FrozenGene — APPROVED** [review]

**Comment by FrozenGene** [comment]
About llvm11, i think we could update it for our ci

**Comment by giuseros** [comment]
Hi @FrozenGene ,
Thanks for approving! Yes, I wanted to do that in a separate PR to not pollute this one. Is that ok with you?

**Comment by FrozenGene** [comment]
> Hi @FrozenGene ,
> Thanks for approving! Yes, I wanted to do that in a separate PR to not pollute this one. Is that ok with you?

yes

**Pull request merged**